### PR TITLE
Register fold import action

### LINKED ISSUE #130678: Add keybindings for import folding
Import folding feature is awesome! But it should have own keybinding to fold / unfold imports.

### PR DESCRIPTION
**Pull request by BABA983**
Close #130678 

https://github.com/user-attachments/assets/0a73fe12-db69-4f33-ad21-998c21ed7561

### CODE REVIEW

**Comment by aeschli** [comment]
The folding model already knows which ranges are import ranges: type === FoldingRangeKind.Imports

See here as an example:
https://github.com/microsoft/vscode/blob/c95571667318a8ff30e27e7302f29cc6318a4001/src/vs/editor/contrib/folding/browser/folding.ts#L877

Of course this only works for languages that have a provider. Your PR also has the same problem.

**Comment by BABA983** [comment]
> The folding model already knows which ranges are import ranges: type === FoldingRangeKind.Imports
> 
> See here as an example:
> 
> https://github.com/microsoft/vscode/blob/c95571667318a8ff30e27e7302f29cc6318a4001/src/vs/editor/contrib/folding/browser/folding.ts#L877
> 
> Of course this only works for languages that have a provider. Your PR also has the same problem.

I have update the code get regions from folding model.

> Of course this only works for languages that have a provider. Your PR also has the same problem.

That seems to make sense to have a provider figure out what range is import range in different language?

**Comment by aeschli** [comment]
You can use the already existing function `setCollapseStateForType`: https://github.com/microsoft/vscode/blob/c95571667318a8ff30e27e7302f29cc6318a4001/src/vs/editor/contrib/folding/browser/foldingModel.ts#L436

Depending on what extensions are installed, some language have rich language support and a folding provider, some don't.
For the languages that do not have a folding provider, folding ranges are computed using the indentation.
That might match the import sections, but must not.

**Comment by BABA983** [comment]
> You can use the already existing function setCollapseStateForType

If we reuse the function we might need to know if the import range is collapsed or expanded.
Maybe we need a new function name `toggleCollapseStateForType`?

> That might match the import sections, but must not.

If I understand right we should toggle fold import when there are provider provided? Like [getFoldingRangeProviders](https://github.com/microsoft/vscode/blob/c95571667318a8ff30e27e7302f29cc6318a4001/src/vs/editor/contrib/folding/browser/folding.ts#L286) here?

**Comment by aeschli** [comment]
> If we reuse the function we might need to know if the import range is collapsed or expanded.
> Maybe we need a new function name toggleCollapseStateForType?

Do you want to toggle or set? I assumed from the name of the command `fold import` it is about setting the state to folded.

The command will just be a no-op if there's fodling provider that knows about imports. That's fine.

**Comment by BABA983** [comment]
> > If we reuse the function we might need to know if the import range is collapsed or expanded.
> > Maybe we need a new function name toggleCollapseStateForType?
> 
> Do you want to toggle or set? I assumed from the name of the command `fold import` it is about setting the state to folded.
> 
> The command will just be a no-op if there's fodling provider that knows about imports. That's fine.

Maybe `toggle` is better? Only fold import seems quite limited. Separating the two commands `fold` and `unfold` seems to give the user more flexibility. What do you think?

**Comment by aeschli** [comment]
Toggle is probably more useful.

File: src/vs/editor/contrib/folding/browser/folding.ts [hunk 1]
```diff
@@ -1231,6 +1231,35 @@ class RemoveFoldRangeFromSelectionAction extends FoldingAction<void> {
...
+
+	async invoke(foldingController: FoldingController, foldingModel: FoldingModel, editor: ICodeEditor, args: void, languageConfigurationService: ILanguageConfigurationService): Promise<void> {
+		const regionsToToggle: FoldingRegion[] = [];
+		const regions = foldingModel.regions;
+		for (let i = regions.length - 1; i >= 0; i--) {
+			if (regions.getType(i) === FoldingRangeKind.Imports.value) {
+				regionsToToggle.push(regions.toRegion(i));
+			}
+		}
+		foldingModel.toggleCollapseState(Array.from(regionsToToggle.values()));
```
Comment (aeschli) [inline src/vs/editor/contrib/folding/browser/folding.ts:1257]: Use `foldingModel.toggleCollapseState(regionsToToggle);`

File: src/vs/editor/contrib/folding/browser/folding.ts [hunk 2]
```diff
@@ -1231,6 +1231,35 @@ class RemoveFoldRangeFromSelectionAction extends FoldingAction<void> {
...
+			alias: 'Toggle Import Fold',
+			precondition: CONTEXT_FOLDING_ENABLED,
+			kbOpts: {
+				kbExpr: EditorContextKeys.editorTextFocus,
+				weight: KeybindingWeight.EditorContrib
+			}
+		});
+	}
+
+	async invoke(foldingController: FoldingController, foldingModel: FoldingModel, editor: ICodeEditor, args: void, languageConfigurationService: ILanguageConfigurationService): Promise<void> {
```
Comment (aeschli) [inline src/vs/editor/contrib/folding/browser/folding.ts:1249]: remove the arguments that are not used

**Review by aeschli — CHANGES_REQUESTED** [review]

**Review by aeschli — APPROVED** [review]

**Comment by aeschli** [comment]
Thanks @BABA983 !

**Review by lszomoru — APPROVED** [review]

**Pull request merged**